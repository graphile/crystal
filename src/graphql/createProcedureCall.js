const createProcedureCall = procedure => {
  // Construct the qualified procedure name.
  const procedureName = `"${procedure.schema.name}"."${procedure.name}"`

  // Construct the argument list for the procedure call.
  const procedureArgsList = Array.from(procedure.args).map(([name, type], i) => {
    const placeholder = `$${i + 1}`

    // If the type of this argument is a table type, we will be expecting JSON
    // but we need that JSON to be a table. Therefore we run it through
    // `json_populate_record`.
    if (type.isTableType)
      return `json_populate_record(null::"${type.table.schema.name}"."${type.table.name}", ${placeholder})`

    return placeholder
  }).join(', ')

  // Add the procedure name with the procedure argument list.
  const procedureCall = `${procedureName}(${procedureArgsList})`

  return procedureCall
}

export default createProcedureCall
