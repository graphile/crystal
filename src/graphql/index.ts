import createGraphQLSchema from './schema/createGQLSchema'

export {
  createGraphQLSchema,
}
