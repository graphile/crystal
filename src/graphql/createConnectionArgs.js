import { memoize, fromPairs, includes, snakeCase, toUpper } from 'lodash'
import { GraphQLEnumType, GraphQLInt, GraphQLBoolean } from 'graphql'
import { CursorType } from './types.js'
import getType from './getType.js'

const createConnectionArgs = (table, ignoreColumnConditions = []) => ({
  // The column specified by `orderBy` means more than just the order to
  // return items in. This column is also the column we will use for
  // cursors.
  orderBy: {
    type: createTableOrderingEnum(table),
    description:
      'The order the resulting items should be returned in. This argument ' +
      'is also important as it is used in creating pagination cursors. This ' +
      'value’s default is the primary key for the object.',
    defaultValue: (() => {
      const column = table.getPrimaryKeys()[0]
      if (column) return column.name
      return null
    })(),
  },
  first: {
    type: GraphQLInt,
    description:
      'The top `n` items in the set to be returned. Can’t be used ' +
      'with `last`.',
  },
  last: {
    type: GraphQLInt,
    description:
      'The bottom `n` items in the set to be returned. Can’t be used ' +
      'with `first`.',
  },
  before: {
    type: CursorType,
    description: 'Constrains the set to nodes *before* this cursor in the specified ordering.',
  },
  after: {
    type: CursorType,
    description: 'Constrains the set to nodes *after* this cursor in the specified ordering.',
  },
  offset: {
    type: GraphQLInt,
    description: 'An integer offset representing how many items to skip in the set.',
  },
  descending: {
    type: GraphQLBoolean,
    description:
      'If `true` the nodes will be in descending order, if `false` the ' +
      'items will be in ascending order. `false` by default.',
    defaultValue: false,
  },
  ...fromPairs(
    table
    .getColumns()
    // If `ignoreColumnConditions` is set to true, all column conditions will
    // be disabled. If `ignoreColumnConditions` is an array, only certain
    // conditions will be ignored.
    .filter(column => (ignoreColumnConditions === true ? false : !includes(ignoreColumnConditions, column)))
    .map(column => [column.getFieldName(), {
      type: getType(column.type),
      description:
        'Filters the resulting set with an equality test on the ' +
        `${column.getMarkdownFieldName()} field.`,
      // TODO: Deprecate this…
      // deprecationReason:
      //   'Simple equality testing is insufficient for the nodes field and just ' +
      //   'adds noise. Instead use procedures for custom set filtering.',
    }])
  ),
})

export default createConnectionArgs

/**
 * Creates an ordering enum which simply contains all of a `Table`s columns.
 *
 * @param {Table} table
 * @returns {GraphQLEnumType}
 */
// TODO: Some way to eliminate some columns from ordering enum?
export const createTableOrderingEnum = memoize(table =>
  new GraphQLEnumType({
    name: `${table.getTypeName()}Ordering`,
    description: `Properties with which ${table.getMarkdownTypeName()} can be ordered.`,

    values: fromPairs(
      table
      .getColumns()
      .map(column => [toUpper(snakeCase(column.getFieldName())), {
        value: column.name,
        description: column.description,
      }])
    ),
  }))
