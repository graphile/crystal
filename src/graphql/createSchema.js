import { GraphQLSchema } from 'graphql'
import createQueryType from './query/createQueryType.js'
import createMutationType from './mutation/createMutationType.js'

/**
 * Creates a GraphQLSchema from a PostgreSQL schema.
 *
 * @param {Schema} schema
 * @returns {GrpahQLSchema}
 */
const createGraphqlSchema = (schema, options) =>
  new GraphQLSchema({
    query: createQueryType(schema, options),
    mutation: createMutationType(schema, options),
  })

export default createGraphqlSchema
