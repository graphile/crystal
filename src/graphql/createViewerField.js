import { memoize } from 'lodash'
import { GraphQLObjectType, GraphQLNonNull } from 'graphql'
import createQueryFields from './query/createQueryFields.js'

const createViewerField = memoize(schema => ({
  type: new GraphQLNonNull(new GraphQLObjectType({
    name: 'Viewer',
    description: '',
    fields: createQueryFields(schema),
  })),

  description:
    'A single entry query for advanced data clients like Relay. Nothing ' +
    'special at all, if you don’t know what this field is for, you probably ' +
    'don’t need it.',

  resolve: source => source || {},
}))

export default createViewerField
