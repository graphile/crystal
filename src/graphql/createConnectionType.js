import { memoize } from 'lodash'
import createTableType from './createTableType.js'
import { PageInfoType, CursorType } from './types.js'

import {
  GraphQLObjectType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLInt,
} from 'graphql'

const createTableConnectionType = memoize(table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Connection`,
    description: `A connection to a list of ${table.getMarkdownTypeName()} items`,

    // TODO: Implement a `ConnectionType` interface

    fields: {
      pageInfo: {
        type: new GraphQLNonNull(PageInfoType),
        description: `Information to aid in pagination of type ${table.getMarkdownTypeName()}.`,
        resolve: pageInfo => pageInfo,
      },
      totalCount: {
        type: GraphQLInt,
        description: 'All of the items available to be queried in this connection.',
        resolve: ({ totalCount }) => totalCount,
      },
      nodes: {
        type: new GraphQLList(createTableType(table)),
        description: `The queried list of ${table.getMarkdownTypeName()}.`,
        resolve: ({ nodes }) => nodes,
      },
      edges: {
        type: new GraphQLList(createTableEdgeType(table)),
        description: 'A single item and a cursor to aid in pagination.',
        resolve: ({ edges }) => edges,
      },
    },
  })
)

export default createTableConnectionType

const createTableEdgeType = table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Edge`,
    description: `An edge in the \`${table.getTypeName()}Connection\`.`,

    fields: {
      cursor: {
        type: new GraphQLNonNull(CursorType),
        description: 'The cursor describing the position of the edgeâ€™s associated node.',
        resolve: ({ cursor }) => cursor || 'null',
      },
      node: {
        type: createTableType(table),
        description: 'The item at the end of the edge.',
        resolve: ({ node }) => node,
      },
    },
  })
