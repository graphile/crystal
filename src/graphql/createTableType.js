import { memoize, fromPairs, camelCase, constant, assign } from 'lodash'
import { GraphQLObjectType, GraphQLID } from 'graphql'
import { $$rowTable } from '../symbols.js'
import { NodeType, toID } from './types.js'
import getColumnType from './getColumnType.js'
import resolveTableSingle from './resolveTableSingle.js'
import createConnectionType from './createConnectionType.js'
import createConnectionArgs from './createConnectionArgs.js'
import resolveConnection from './resolveConnection.js'
import createProcedureReturnType from './createProcedureReturnType.js'
import createProcedureArgs from './createProcedureArgs.js'
import resolveProcedure from './resolveProcedure.js'
import createProcedureCall from './createProcedureCall.js'

/**
 * Creates the `GraphQLObjectType` for a table.
 *
 * This function is memoized because it will be called often for the same
 * table. Think that both the single and list fields need an instance of the
 * table type. Instead of passing the table type around as a parameter, it is
 * more functional to just memoize this function.
 *
 * @param {Client} client
 * @param {Table} table
 * @returns {GraphQLObjectType}
 */
const createTableType = memoize(table => {
  const columns = table.getColumns()

  // If we have no fields, GraphQL will be sad. Make sure we have meaningful
  // fields by erroring if there are no columns.
  if (columns.length === 0) {
    throw new Error(
      `PostgreSQL schema '${table.schema.name}' contains table '${table.name}' ` +
      'which does not have any columns. To generate a GraphQL schema all ' +
      'tables must have at least one column.'
    )
  }

  const primaryKeys = table.getPrimaryKeys()
  const isNode = primaryKeys.length !== 0

  return new GraphQLObjectType({
    // Creates a new type where the name is a PascalCase version of the table
    // name and the description is the associated comment in PostgreSQL.
    name: table.getTypeName(),
    description: table.description,

    // If the table has no primary keys, it shouldn’t implement `Node`.
    interfaces: isNode ? [NodeType] : [],

    isTypeOf: value => value[$$rowTable] === table,

    // Make sure all of our columns have a corresponding field. This is a thunk
    // because `createForeignKeyField` may have a circular dependency.
    fields: () => ({
      // If the table is a node, add the `id` field. Otherwise don’t add
      // anything.
      ...(isNode ? {
        id: {
          type: GraphQLID,
          description: `The globally unique identifier for this ${table.getMarkdownTypeName()}.`,
          resolve: source => toID(table.name, primaryKeys.map(column => source[column.name])),
        },
      } : {}),
      // Add the column fields.
      ...fromPairs(
        columns.map(column => [column.getFieldName(), createColumnField(column)])
      ),
      // Add the computed column fields.
      ...fromPairs(
        table.getComputedColumns().map(procedure => [
          procedure.getFieldName(table.name),
          createProcedureField(procedure),
        ])
      ),
      // Add foreign key field references.
      ...fromPairs(
        table.getForeignKeys().map(foreignKey => {
          const columnNames = foreignKey.nativeColumns.map(({ name }) => name)
          const name = `${foreignKey.foreignTable.name}_by_${columnNames.join('_and_')}`
          return [camelCase(name), createForeignKeyField(foreignKey)]
        })
      ),
      // Add reverse foreign key field references.
      ...fromPairs(
        table.getReverseForeignKeys().map(foreignKey => {
          const columnNames = foreignKey.nativeColumns.map(({ name }) => name)
          const name = `${foreignKey.nativeTable.name}_nodes_by_${columnNames.join('_and_')}`
          return [camelCase(name), createForeignKeyReverseField(foreignKey)]
        })
      ),
    }),
  })
})

export default createTableType

const createColumnField = column => ({
  type: getColumnType(column),
  description: column.description,
  resolve: source => source[column.name],
})

const createForeignKeyField = ({ nativeTable, nativeColumns, foreignTable, foreignColumns }) => ({
  type: createTableType(foreignTable),
  description:
    `Queries a single ${foreignTable.getMarkdownTypeName()} node related to ` +
    `the ${nativeTable.getMarkdownTypeName()} type.`,

  resolve: resolveTableSingle(
    foreignTable,
    foreignColumns,
    source => nativeColumns.map(({ name }) => source[name])
  ),
})

const createForeignKeyReverseField = ({ nativeTable, nativeColumns, foreignTable, foreignColumns }) => ({
  type: createConnectionType(nativeTable),
  description:
    `Queries and returns a set of ${nativeTable.getMarkdownTypeName()} ` +
    `nodes that are related to the ${foreignTable.getMarkdownTypeName()} source ` +
    'node.',

  args: createConnectionArgs(nativeTable, nativeColumns),

  resolve: resolveConnection(
    nativeTable,
    source => fromPairs(foreignColumns.map(({ name }, i) => [nativeColumns[i].name, source[name]]))
  ),
})

const createProcedureField = procedure => {
  const [tableArg, ...argEntries] = Array.from(procedure.args)
  const [tableArgName, tableArgType] = tableArg
  const returnTable = procedure.getReturnTable()

  const procedureArgs = createProcedureArgs(
    procedure,
    (argName, argType) => argName === tableArgName && argType === tableArgType,
  )

  // If this is a connection, return a completely different field…
  if (procedure.returnsSet && returnTable) {
    return {
      type: createConnectionType(returnTable),
      description: procedure.description,

      args: {
        // Add the arguments for the procedure…
        ...procedureArgs,
        // Add the arguments for connections…
        ...createConnectionArgs(returnTable, true),
      },

      // Resolve the connection.
      resolve: resolveConnection(
        returnTable,
        constant({}),
        (source, args) => ({
          // Use the text from the procedure call.
          text: createProcedureCall(procedure),
          // Use values from argument entries and the table source value.
          values: [source, ...argEntries.map(([name]) => args[camelCase(name)])],
        }),
      ),
    }
  }

  return {
    type: createProcedureReturnType(procedure),
    description: procedure.description,

    // Create the arguments and omit the table argument.
    args: procedureArgs,

    // Resolve the procedure, using the source row as the argument we omit.
    resolve: resolveProcedure(
      procedure,
      (source, args) => assign(args, { [camelCase(tableArgName)]: source }),
    ),
  }
}
