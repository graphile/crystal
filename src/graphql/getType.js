import {
  GraphQLBoolean,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLList,
  GraphQLEnumType,
} from 'graphql'

import { types } from 'pg'
import { memoize, fromPairs, upperFirst, camelCase, snakeCase, toUpper } from 'lodash'
import createTableType from './createTableType.js'

import {
  DateType,
  BigIntType,
  PointType,
  CircleType,
  IntervalType,
  JSONType,
  UUIDType,
} from './types.js'

/**
 * A map of types where the PostgreSQL type is the key and the GraphQL type is
 * the value.
 *
 * In order to see available PostgreSQL types in your database run the
 * following query:
 *
 * ```sql
 * select typname, oid, typarray from pg_type where typtype = 'b' order by oid;
 * ```
 *
 * Also see the [`pg-types`][1] module’s source code to see what types they
 * are parsing.
 *
 * [1]: https://github.com/brianc/node-pg-types
 */
const postgresToGraphQLTypes = new Map([
  [20, BigIntType],
  [21, GraphQLInt],
  [23, GraphQLInt],
  [26, GraphQLInt],
  [700, GraphQLFloat],
  [701, GraphQLFloat],
  [16, GraphQLBoolean],
  [1082, DateType],
  [1114, DateType],
  [1184, DateType],
  [600, PointType],
  [718, CircleType],
  [1000, new GraphQLList(GraphQLBoolean)],
  [1005, new GraphQLList(GraphQLInt)],
  [1007, new GraphQLList(GraphQLInt)],
  [1028, new GraphQLList(GraphQLInt)],
  [1016, new GraphQLList(BigIntType)],
  [1021, new GraphQLList(GraphQLFloat)],
  [1022, new GraphQLList(GraphQLFloat)],
  [1231, new GraphQLList(GraphQLFloat)],
  [1014, new GraphQLList(GraphQLString)],
  [1015, new GraphQLList(GraphQLString)],
  [1008, new GraphQLList(GraphQLString)],
  [1009, new GraphQLList(GraphQLString)],
  [1115, new GraphQLList(DateType)],
  [1182, new GraphQLList(DateType)],
  [1185, new GraphQLList(DateType)],
  [1186, IntervalType],
  [114, JSONType],
  [3802, JSONType],
  [199, new GraphQLList(JSONType)],
  [3807, new GraphQLList(JSONType)],
  [2951, new GraphQLList(GraphQLString)],
  [791, new GraphQLList(GraphQLString)],
  [1183, new GraphQLList(GraphQLString)],
  [1700, GraphQLFloat],
  [2950, UUIDType],
  [18, GraphQLString],
  [25, GraphQLString],
  [1043, GraphQLString],
])

// Override custom type parsers.
// TODO: This is a temporary solution.
types.setTypeParser(600, String)
types.setTypeParser(718, String)
types.setTypeParser(1186, String)

/**
 * Gets a GraphQL type for a PostgreSQL type.
 *
 * @param {Column} column
 * @returns {GraphQLType}
 */
const getType = memoize(type => {
  // If the type is an enum, let’s create an enum for it.
  if (type.isEnum) {
    return new GraphQLEnumType({
      name: upperFirst(camelCase(type.name)),
      description: type.description,
      values: fromPairs(
        type.variants
        .map(variant => [toUpper(snakeCase(variant)), { value: variant }])
      ),
    })
  }

  // If the type is a domain, return the underlying type
  if (type.isDomain)
    return getType(type.baseType)

  // If this type is a table type, use the PostGraphQL table type.
  if (type.isTableType)
    return createTableType(type.table)

  // Return internal type or a string.
  return postgresToGraphQLTypes.get(type.id) || GraphQLString
})

export default getType
