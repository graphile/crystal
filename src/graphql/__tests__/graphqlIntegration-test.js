import { printSchema } from 'graphql'
import forumInventory from './fixtures/forumInventory'
import createGQLSchema from '../schema/createGQLSchema'

test('will generate the correct forum schema', () => {
  const gqlSchema = createGQLSchema(forumInventory)
  const printedSchema = printSchema(gqlSchema)
  expect(printedSchema).toMatchSnapshot()
})
