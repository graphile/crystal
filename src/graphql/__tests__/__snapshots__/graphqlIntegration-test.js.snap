exports[`test will generate the correct forum schema 1`] = `
"# All input for the \`createPerson\` mutation.
input CreatePersonInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Person\` to be created by this mutation.
  person: PersonInput!
}

# The output of our \`createPerson\` mutation.
type CreatePersonPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String

  # The \`Person\` that was created by this mutation.
  person: Person

  # An edge for our \`Person\`. May be used by Relay 1.
  personEdge(
    # The method to use when ordering \`Person\`.
    orderBy: PeopleOrderBy = NAME_ASC
  ): PeopleEdge

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}

# All input for the \`createPost\` mutation.
input CreatePostInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Post\` to be created by this mutation.
  post: PostInput!
}

# The output of our \`createPost\` mutation.
type CreatePostPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String

  # The \`Post\` that was created by this mutation.
  post: Post

  # An edge for our \`Post\`. May be used by Relay 1.
  postEdge(
    # The method to use when ordering \`Post\`.
    orderBy: PostsOrderBy = HEADLINE_ASC
  ): PostsEdge

  # Reads a single \`Person\` that is related to this \`Post\`.
  personByAuthor: Person

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}

# A location in a connection that can be used for resuming pagination.
scalar Cursor

# All input for the \`deletePersonById\` mutation.
input DeletePersonByIdInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Int\` to use when reading a single value.
  id: Int!
}

# All input for the \`deletePerson\` mutation.
input DeletePersonInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The globally unique \`ID\` which will identify a single \`Person\` to be deleted.
  nodeId: ID!
}

# The output of our \`deletePerson\` mutation.
type DeletePersonPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String
  person: Person
  deletedPersonId: ID

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}

# All input for the \`deletePostById\` mutation.
input DeletePostByIdInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Int\` to use when reading a single value.
  id: Int!
}

# All input for the \`deletePost\` mutation.
input DeletePostInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The globally unique \`ID\` which will identify a single \`Post\` to be deleted.
  nodeId: ID!
}

# The output of our \`deletePost\` mutation.
type DeletePostPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String
  post: Post
  deletedPostId: ID

  # Reads a single \`Person\` that is related to this \`Post\`.
  personByAuthor: Person

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}

scalar Email

# The root mutation type which contains root level fields which mutate data.
type Mutation {
  # Creates a single \`Person\`.
  createPerson(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: CreatePersonInput!
  ): CreatePersonPayload

  # Updates a single \`Person\` using its globally unique id and a patch.
  updatePerson(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: UpdatePersonInput!
  ): UpdatePersonPayload

  # Updates a single \`Person\` using a unique key and a patch.
  updatePersonById(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: UpdatePersonByIdInput!
  ): UpdatePersonPayload

  # Updates a single \`Person\` using a unique key and a patch.
  updatePersonByEmail(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: UpdatePersonByEmailInput!
  ): UpdatePersonPayload

  # Deletes a single \`Person\` using its globally unique id.
  deletePerson(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: DeletePersonInput!
  ): DeletePersonPayload

  # Deletes a single \`Person\` using a unique key.
  deletePersonById(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: DeletePersonByIdInput!
  ): DeletePersonPayload

  # Creates a single \`Post\`.
  createPost(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: CreatePostInput!
  ): CreatePostPayload

  # Updates a single \`Post\` using its globally unique id and a patch.
  updatePost(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: UpdatePostInput!
  ): UpdatePostPayload

  # Updates a single \`Post\` using a unique key and a patch.
  updatePostById(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: UpdatePostByIdInput!
  ): UpdatePostPayload

  # Deletes a single \`Post\` using its globally unique id.
  deletePost(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: DeletePostInput!
  ): DeletePostPayload

  # Deletes a single \`Post\` using a unique key.
  deletePostById(
    # The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    input: DeletePostByIdInput!
  ): DeletePostPayload
}

# An object with a globally unique \`ID\`.
interface Node {
  # A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  nodeId: ID!
}

# Information about pagination in a connection.
type PageInfo {
  # When paginating forwards, are there more items?
  hasNextPage: Boolean!

  # When paginating backwards, are there more items?
  hasPreviousPage: Boolean!

  # When paginating backwards, the cursor to continue.
  startCursor: Cursor

  # When paginating forwards, the cursor to continue.
  endCursor: Cursor
}

# A connection to a list of \`Person\` values.
type PeopleConnection {
  # Information to aid in pagination.
  pageInfo: PageInfo!

  # The count of *all* \`Person\` you could get from the connection.
  totalCount: Int

  # A list of edges which contains the \`Person\` and cursor to aid in pagination.
  edges: [PeopleEdge]

  # A list of \`Person\` objects.
  nodes: [Person!]
}

# A \`Person\` edge in the connection.
type PeopleEdge {
  # A cursor for use in pagination.
  cursor: Cursor

  # The \`Person\` at the end of the edge.
  node: Person!
}

# Methods to use when ordering \`Person\`.
enum PeopleOrderBy {
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  FIRST_NAME_ASC
  FIRST_NAME_DESC
  LAST_NAME_ASC
  LAST_NAME_DESC
}

type Person implements Node {
  # A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  nodeId: ID!
  id: Int!
  name: String!
  email: Email!
  firstName: String
  lastName: String
  about: String
}

# A condition to be used against \`Person\` object types. All fields are tested for equality and combined with a logical ‘and.’
input PersonCondition {
  # Checks for equality with the object’s \`id\` field.
  id: Int

  # Checks for equality with the object’s \`name\` field.
  name: String

  # Checks for equality with the object’s \`email\` field.
  email: Email

  # Checks for equality with the object’s \`firstName\` field.
  firstName: String

  # Checks for equality with the object’s \`lastName\` field.
  lastName: String

  # Checks for equality with the object’s \`about\` field.
  about: String
}

input PersonInput {
  id: Int!
  name: String!
  email: Email!
  firstName: String
  lastName: String
  about: String
}

# Represents an update to a \`Person\`. Fields that are set will be updated.
input PersonPatch {
  id: Int
  name: String
  email: Email
  firstName: String
  lastName: String
  about: String
}

type Post implements Node {
  # A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  nodeId: ID!
  id: Int!
  authorId: Int!
  status: PostStatus!
  headline: String!
  body: String

  # Reads a single \`Person\` that is related to this \`Post\`.
  personByAuthor: Person
}

# A condition to be used against \`Post\` object types. All fields are tested for equality and combined with a logical ‘and.’
input PostCondition {
  # Checks for equality with the object’s \`id\` field.
  id: Int

  # Checks for equality with the object’s \`authorId\` field.
  authorId: Int

  # Checks for equality with the object’s \`status\` field.
  status: PostStatus

  # Checks for equality with the object’s \`headline\` field.
  headline: String

  # Checks for equality with the object’s \`body\` field.
  body: String
}

input PostInput {
  id: Int!
  authorId: Int!
  status: PostStatus!
  headline: String!
  body: String
}

# Represents an update to a \`Post\`. Fields that are set will be updated.
input PostPatch {
  id: Int
  authorId: Int
  status: PostStatus
  headline: String
  body: String
}

# A connection to a list of \`Post\` values.
type PostsConnection {
  # Information to aid in pagination.
  pageInfo: PageInfo!

  # The count of *all* \`Post\` you could get from the connection.
  totalCount: Int

  # A list of edges which contains the \`Post\` and cursor to aid in pagination.
  edges: [PostsEdge]

  # A list of \`Post\` objects.
  nodes: [Post!]
}

# A \`Post\` edge in the connection.
type PostsEdge {
  # A cursor for use in pagination.
  cursor: Cursor

  # The \`Post\` at the end of the edge.
  node: Post!
}

# Methods to use when ordering \`Post\`.
enum PostsOrderBy {
  ID_ASC
  ID_DESC
  AUTHOR_ID_ASC
  AUTHOR_ID_DESC
  STATUS_ASC
  STATUS_DESC
  HEADLINE_ASC
  HEADLINE_DESC
}

enum PostStatus {
  UNPUBLISHED
  PUBLISHED
}

# The root query type which gives access points into the data universe.
type Query implements Node {
  # Fetches an object given its globally unique \`ID\`.
  node(
    # The globally unique \`ID\`.
    nodeId: ID!
  ): Node

  # Reads and enables pagination through a set of \`Person\`.
  allPeople(
    # The method to use when ordering \`Person\`.
    orderBy: PeopleOrderBy = NAME_ASC

    # Read all values in the set before (above) this cursor.
    before: Cursor

    # Read all values in the set after (below) this cursor.
    after: Cursor

    # Only read the first \`n\` values of the set.
    first: Int

    # Only read the last \`n\` values of the set.
    last: Int

    # Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    # based pagination. May not be used with \`last\`.
    offset: Int

    # A condition to be used in determining which values should be returned by the collection.
    condition: PersonCondition
  ): PeopleConnection

  # Reads a single \`Person\` using its globally unique \`ID\`.
  person(
    # The globally unique \`ID\` to be used in selecting a single \`Person\`.
    nodeId: ID!
  ): Person
  personById(
    # The \`Int\` to use when reading a single value.
    id: Int!
  ): Person
  personByName(
    # The \`String\` to use when reading a single value.
    name: String!
  ): Person
  personByEmail(
    # The \`String\` to use when reading a single value.
    email: String!
  ): Person

  # Reads and enables pagination through a set of \`Post\`.
  allPosts(
    # The method to use when ordering \`Post\`.
    orderBy: PostsOrderBy = HEADLINE_ASC

    # Read all values in the set before (above) this cursor.
    before: Cursor

    # Read all values in the set after (below) this cursor.
    after: Cursor

    # Only read the first \`n\` values of the set.
    first: Int

    # Only read the last \`n\` values of the set.
    last: Int

    # Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    # based pagination. May not be used with \`last\`.
    offset: Int

    # A condition to be used in determining which values should be returned by the collection.
    condition: PostCondition
  ): PostsConnection

  # Reads a single \`Post\` using its globally unique \`ID\`.
  post(
    # The globally unique \`ID\` to be used in selecting a single \`Post\`.
    nodeId: ID!
  ): Post
  postById(
    # The \`Int\` to use when reading a single value.
    id: Int!
  ): Post

  # Exposes the root query type nested one level down. This is helpful for Relay 1
  # which can only query top level fields if they are in a particular form.
  query: Query!

  # The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  nodeId: ID!
}

# All input for the \`updatePersonByEmail\` mutation.
input UpdatePersonByEmailInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`String\` to use when reading a single value.
  email: String!

  # An object where the defined keys will be set on the \`Person\` identified by our unique key.
  personPatch: PersonPatch!
}

# All input for the \`updatePersonById\` mutation.
input UpdatePersonByIdInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Int\` to use when reading a single value.
  id: Int!

  # An object where the defined keys will be set on the \`Person\` identified by our unique key.
  personPatch: PersonPatch!
}

# All input for the \`updatePerson\` mutation.
input UpdatePersonInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The globally unique \`ID\` which will identify a single \`Person\` to be updated.
  nodeId: ID!

  # An object where the defined keys will be set on the \`Person\` identified by our globally unique \`ID\`.
  personPatch: PersonPatch!
}

# The output of our \`updatePerson\` mutation.
type UpdatePersonPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String
  person: Person

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}

# All input for the \`updatePostById\` mutation.
input UpdatePostByIdInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The \`Int\` to use when reading a single value.
  id: Int!

  # An object where the defined keys will be set on the \`Post\` identified by our unique key.
  postPatch: PostPatch!
}

# All input for the \`updatePost\` mutation.
input UpdatePostInput {
  # An arbitrary string value with no semantic meaning. Will be included in the
  # payload verbatim. May be used to track mutations by the client.
  clientMutationId: String

  # The globally unique \`ID\` which will identify a single \`Post\` to be updated.
  nodeId: ID!

  # An object where the defined keys will be set on the \`Post\` identified by our globally unique \`ID\`.
  postPatch: PostPatch!
}

# The output of our \`updatePost\` mutation.
type UpdatePostPayload {
  # The exact same \`clientMutationId\` that was provided in the mutation input,
  # unchanged and unused. May be used by a client to track mutations.
  clientMutationId: String
  post: Post

  # Reads a single \`Person\` that is related to this \`Post\`.
  personByAuthor: Person

  # Our root query field type. Allows us to run any query from our mutation payload.
  query: Query
}
"
`;
